### ISSUE TITLE
Tracking Issue for `-Zgit`

### ISSUE THREAD

**Issue by weihanglo** [rust-lang/cargo#13285]
### Summary

Original issue: #1171. See also #11813
Implementation: #13252
Documentation: https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#my-feature

`-Zgit` is a collection of new Git features in Cargo. Currently we have experiments on

* Shallow clones for git dependencies
* Shallow clones for registry indexes

### Unresolved Issues

* [ ] No native support of local `file` shallow clones for either gitoxide or libgit2
  * https://github.com/Byron/gitoxide/issues/734
  * https://github.com/libgit2/libgit2/issues/6634
* [ ] Add tests for git2 shallow clones. This is blocked on the support of `file:` protocol.
* [ ] Is `*-shallow` directory layout optimal? Can we avoid that but remain compatible?

### Future Extensions

See "Future Extensions" in #11813. Most of them depend on upstream features in libgit2.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

**Comment by NoelJacob** [comment]
My `~/.cargo/config.toml` looks like this:
```
git.shallow-index = true
git.shallow-deps = true
gitoxide.fetch = true
gitoxide.checkout = true
```
But on running `cargo help` I get :
```
>  cargo help
error: missing field `internal_use_git2`

>  cargo help
warning: unused config key `unstable.git.shallow-index` in `/home/noel/.cargo/config.toml`
warning: unused config key `unstable.git.shallow-deps` in `/home/noel/.cargo/config.toml`
error: missing field `shallow_index`
```
Keep repeating `cargo <command>` and the error seems to repeat between the two. Passing `-Zgit -Zgitoxide` seems to not give error.

**Comment by weihanglo** [comment]
Under `~/.cargo/git`, check if the git clone contains a `-shallow` suffix.

```console
~/.cargo/git $ tree -L 2
./
├── checkouts/
│  └── empty-library-0161899a0cd3bb43-shallow/
└── db/
   └── empty-library-0161899a0cd3bb43-shallow/
```

As of this writing, you need this in your `.cargo/config.toml`:

```toml
[unstable.git]
shallow_index = true
shallow_deps = true
[unstable.gitoxide]
fetch = true
checkout = true
list_files = true
internal_use_git2 = false
```

It is not ergonomic, but while it is still unstable, I am not too concerned. If people want to fix it, go for it! See <https://github.com/rust-lang/cargo/issues/11813#issuecomment-1817517629> as well.

**Comment by linyihai** [comment]
(Maybe a bit off topic)  what is the relationship between `-Zgit` and `-Zgitoxide`? I see both features being developed at the same time, will they be stable in the future?

**Comment by weihanglo** [comment]
`-Zgitoxide` was first developed. The shallow clone support in libgit2 came out after and then we decided to split into `-Zgit` and `-Zgitoxide`. The former controls Git related unstable features in general, and with the latter some of those operations will be performed by gitoxide.

**Comment by soloturn** [comment]
for build environments partial clones in the sense of 
```
--filter=tree:0
```
make most sense. shallow clones i the sense of ```-depth=1``` may have undesired side effects. see e.g. here for details: https://github.blog/2020-12-21-get-up-to-speed-with-partial-clone-and-shallow-clone/

**Comment by weihanglo** [comment]
Sparse checkout and partial clone have been discussed in <https://github.com/rust-lang/cargo/issues/11165>.

It is possible that Cargo ditch or adopt anything for `-Zgit` since it is still an experimental unstable feature. See Byron's comment <https://github.com/rust-lang/cargo/issues/11165#issuecomment-2029220994> for some integration challenges. I can see that download blobs on demand would be helpful for fetching registry index. I am not sure about git dependencies. People might want a "fetch-first-then-offline-build" environment, and we need to model a lazy mechanism for it in Cargo, which is also a bit challenging.

**Comment by weihanglo** [comment]
About this unresolved question:

> Is `*-shallow` directory layout optimal? Can we avoid that but remain compatible?

If we're going to do an incompatible cache layout change, we may want to consider switching to bare repos: https://github.com/rust-lang/cargo/issues/9503

**Comment by weihanglo** [comment]
>  No native support of local file shallow clones for either gitoxide or libgit2 

If we want to stabilize this soon, should we mark the lack of `file` protocol support as a caveat. Local file protocol makes less sense for shallow clones IMO.

**Comment by Logarithmus** [comment]
@weihanglo Hello, thank you for the feature, I've placed the following lines in my global cargo config:

```toml
[net]
git-fetch-with-cli = true

[unstable]
git = true
```

I have to use `git-fetch-with-cli` in order for GitLab private repo auth to work properly.
In my crate I depend on another crate as a git dependency:

```toml
[dependencies]
some-crate = { git = "https://sample.com/some-repo.git", branch = "main" }
```

Now `cargo` creates subdirectories with `-shallow` suffix in `git/checkouts` and `git/db`. However, if I `cd` into one of those subdirs, e.g. `$CARGO_HOME/git/checkouts/some-repo-64ba54c0d7bada33-shallow/7e1f696`, and run `git log`, I can see entire commit history for the branch!

Well, that's an improvement over default behaviour of fetching all branches and tags of the repo, but still, I'd like it to use `--depth 0` or something like that and only fetch the last commit from the branch!

The thing is that in the past, `some-crate` from `some-repo` used to use `cargo vendor` and contained huge `vendor` directory. It's not the case anymore, but `vendor/` directory is still left in git commit history. So when `cargo` fetches this repo's current state, it ends up fetching huge git pack objects weighing tens of megabytes...

And in fact I depend on multiple such crates, so it downloads this trash for each of them...

**Comment by weihanglo** [comment]
@Logarithmus thanks for the feedback!

That might be a bug. Would be appreciated if you dont mind opening a new issue with a minimal reproducible example, so that we can track the fix of it separately :)

That said, I couldn't reproduce your case with the follow steps:

```bash
cargo new foo --lib
cd foo
mkdir -p .cargo

cat > .cargo/config.toml <<'EOF'
[net]
git-fetch-with-cli = true
[unstable]
git = true
EOF

CARGO_HOME=cargo-home cargo +nightly add --git https://github.com/dtolnay/unicode-ident unicode-ident --branch master --verbose
#     Updating git repository `https://github.com/dtolnay/unicode-ident`
#      Running `git fetch --no-tags --depth=1 --verbose --force --update-head-ok 'https://github.com/dtolnay/unicode-ident' '+refs/heads/master:refs/remotes/origin/master'`
# POST git-upload-pack (364 bytes)
# POST git-upload-pack (171 bytes)
# remote: Enumerating objects: 50, done.
# remote: Counting objects: 100% (50/50), done.
# remote: Compressing objects: 100% (39/39), done.
# remote: Total 50 (delta 4), reused 30 (delta 1), pack-reused 0 (from 0)
# Unpacking objects: 100% (50/50), 49.68 KiB | 959.00 KiB/s, done.
# From https://github.com/dtolnay/unicode-ident
#  * [new branch]      master     -> origin/master
#       Adding unicode-ident (git) to dependencies
#     Updating git repository `https://github.com/dtolnay/unicode-ident`
#      Locking 1 package to latest Rust 1.93.0-nightly compatible version
#       Adding unicode-ident v1.0.22 (https://github.com/dtolnay/unicode-ident?branch=master#b5db13da)


cd cargo-home/git/checkouts/unicode-ident-60cd364b05b44f8c-shallow/b5db13d
git rev-list --count --all
# 1 (only b5db13dab47e9b635cfef11d07ea255b7b1203b9)
```

You could put `--verbose` in `cargo fetch` and check if the git CLI execution has  the `--depth=1` flag.
